pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,229 @@
+from gcloud.aio.auth import BUILD_GCLOUD_REST
+
+if BUILD_GCLOUD_REST:
+    pass
+else:
+    import asyncio
+    import logging
+    import time
+    from typing import Awaitable
+    from typing import Callable
+    from typing import List
+    from typing import Optional
+    from typing import Tuple
+    from typing import TYPE_CHECKING
+
+    import aiohttp
+    from gcloud.aio.pubsub.subscriber_client import SubscriberClient
+    from gcloud.aio.pubsub.subscriber_message import SubscriberMessage
+    from gcloud.aio.pubsub.metrics_agent import MetricsAgent
+
+    log = logging.getLogger(__name__)
+
+    if TYPE_CHECKING:
+        MessageQueue = asyncio.Queue[Tuple[SubscriberMessage,  # pylint: disable=unsubscriptable-object
+                                           float]]
+    else:
+        MessageQueue = asyncio.Queue
+    ApplicationHandler = Callable[[SubscriberMessage], Awaitable[None]]
+
+    class AckDeadlineCache:
+        def __init__(self, subscriber_client: SubscriberClient,
+                     subscription: str, cache_timout: int):
+            self.subscriber_client = subscriber_client
+            self.subscription = subscription
+            self.cache_timeout = cache_timout
+            self.ack_deadline: float = float('inf')
+            self.last_refresh: float = 0
+
+        async def get(self) -> float:
+            if self.ack_deadline == float('inf') or self.cache_outdated():
+                await self.refresh()
+            return self.ack_deadline
+
+        async def refresh(self) -> None:
+            try:
+                sub = await self.subscriber_client.get_subscription(
+                    self.subscription)
+                self.ack_deadline = float(sub['ackDeadlineSeconds'])
+            except Exception as e:
+                log.warning(
+                    'Failed to refresh ackDeadlineSeconds value', exc_info=e)
+            self.last_refresh = time.perf_counter()
+
+        def cache_outdated(self) -> bool:
+            if (time.perf_counter() - self.last_refresh) > self.cache_timeout:
+                return True
+            return False
+
+    async def acker(subscription: str,
+                    ack_queue: 'asyncio.Queue[str]',
+                    subscriber_client: 'SubscriberClient',
+                    ack_window: float,
+                    metrics_client: Optional[MetricsAgent]) -> None:
+        ack_ids: List[str] = []
```

Comment:
nit: initialize this at the beginning of each loop rather than here as well as at the end of the loop, to avoid doing it twice (where it could eventually out of sync) and to make it more clear that this value is meant to be temporary per-loop.

Reply:
OOoh, that's the piece you were missing in `ack_ids` thread.